crates/task-impls/src/quorum_vote/drb_computations.rs

```diff
@@ -0,0 +1,158 @@
+
+    /// Stored inputs to computations
+    seeds: BTreeMap<TYPES::Epoch, DrbSeedInput>,
+}
+
+impl<TYPES: NodeType> DrbComputations<TYPES> {
+    #[must_use]
+    /// Create a new DrbComputations
+    pub fn new() -> Self {
+        Self {
+            results: BTreeMap::new(),
+            task: None,
+            seeds: BTreeMap::new(),
+        }
+    }
+
+    /// If a task is currently live AND has finished, join it and save the result.
+    /// If a task is currently live and NOT finished, abort it.
+    /// self.task will always be None after this call.
+    async fn join_or_abort_task(&mut self) {
+        if let Some((epoch, join_handle)) = &mut self.task {
+            if join_handle.is_finished() {
+                match join_handle.await {
+                    Ok(result) => {
+                        self.results.insert(*epoch, result);
+                    }
+                    Err(e) => {
+                        tracing::error!("error joining DRB computation task: {e:?}");
+                    }
+                }
+            } else {
+                join_handle.abort();
+            }
+            self.task = None;
+        }
+    }
+
+    /// If a task is currently live AND has finished, join it and save the result.
+    /// If the epoch for the calculation was the same as the provided epoch, return true
+    /// If a task is currently live and NOT finished, abort it UNLESS the task epoch is the same as
+    /// cur_epoch, in which case keep letting it run and return true.
+    /// Return false if a task should be spawned for the given epoch.
+    async fn join_or_abort_task_for_epoch(&mut self, epoch: TYPES::Epoch) -> bool {
+        if let Some((task_epoch, join_handle)) = &mut self.task {
+            if join_handle.is_finished() {
+                match join_handle.await {
+                    Ok(result) => {
+                        self.results.insert(*task_epoch, result);
+                        let result = *task_epoch == epoch;
+                        self.task = None;
+                        result
+                    }
+                    Err(e) => {
+                        tracing::error!("error joining DRB computation task: {e:?}");
+                        false
+                    }
+                }
+            } else if *task_epoch == epoch {
+                true
+            } else {
+                join_handle.abort();
+                self.task = None;
+                false
+            }
+        } else {
+            false
+        }
+    }
+
+    /// Stores a seed for a particular epoch for later use by start_new_task()
+    pub fn store_seed(&mut self, epoch: TYPES::Epoch, drb_seed_input: DrbSeedInput) {
+        self.seeds.insert(epoch, drb_seed_input);
+    }
+
+    /// Starts a new task. If a previous task is currently live, will attempt to join it (if finished), or abort it.
+    pub async fn start_new_task(&mut self, epoch: TYPES::Epoch) {
```

Comment:
Looks like we'll always use `start_task_if_not_running` rather than this function, so maybe we can remove it. Same for `join_or_abort_task`.